backend/internetarchive/internetarchive.go

```diff
@@ -0,0 +1,1086 @@
+	FrontEndpoint   string               `config:"front_endpoint"`
+	DisableChecksum bool                 `config:"disable_checksum"`
+	WaitArchive     bool                 `config:"wait_archive"`
+	Enc             encoder.MultiEncoder `config:"encoding"`
+}
+
+// Fs represents an IAS3 remote
+type Fs struct {
+	name     string       // name of this remote
+	root     string       // the path we are working on if any
+	opt      Options      // parsed config options
+	features *fs.Features // optional features
+	srv      *rest.Client // the connection to IAS3
+	front    *rest.Client // the connection to frontend
+	pacer    *fs.Pacer    // pacer for API calls
+	ctx      context.Context
+}
+
+// Object describes a file at IA
+type Object struct {
+	fs      *Fs       // reference to Fs
+	remote  string    // the remote path
+	modTime time.Time // last modified time
+	size    int64     // size of the file in bytes
+	md5     string    // md5 hash of the file presented by the server
+	sha1    string    // sha1 hash of the file presented by the server
+	crc32   string    // crc32 of the file presented by the server
+}
+
+// IAFile reprensents a subset of object in MetadataResponse.Files
+type IAFile struct {
+	Name string `json:"name"`
+	// Source     string `json:"source"`
+	Mtime       string          `json:"mtime"`
+	RcloneMtime json.RawMessage `json:"rclone-mtime"`
+	Size        string          `json:"size"`
+	Md5         string          `json:"md5"`
+	Crc32       string          `json:"crc32"`
+	Sha1        string          `json:"sha1"`
+}
+
+// MetadataResponse reprensents subset of the JSON object returned by (frontend)/metadata/
+type MetadataResponse struct {
+	Files    []IAFile `json:"files"`
+	ItemSize int64    `json:"item_size"`
+}
+
+// ModMetadataResponse represents response for amending metadata
+type ModMetadataResponse struct {
+	// https://archive.org/services/docs/api/md-write.html#example
+	Success bool   `json:"success"`
+	Error   string `json:"error"`
+}
+
+// Name of the remote (as passed into NewFs)
+func (f *Fs) Name() string {
+	return f.name
+}
+
+// Root of the remote (as passed into NewFs)
+func (f *Fs) Root() string {
+	return f.root
+}
+
+// String converts this Fs to a string
+func (f *Fs) String() string {
+	bucket, file := f.split("")
+	if bucket == "" {
+		return "Internet Archive root"
+	}
+	if file == "" {
+		return fmt.Sprintf("Internet Archive item %s", bucket)
+	}
+	return fmt.Sprintf("Internet Archive item %s path %s", bucket, file)
+}
+
+// Features returns the optional features of this Fs
+func (f *Fs) Features() *fs.Features {
+	return f.features
+}
+
+// Hashes returns type of hashes supported by IA
+func (f *Fs) Hashes() hash.Set {
+	return hash.NewHashSet(hash.MD5, hash.SHA1, hash.CRC32)
+}
+
+// Precision returns the precision of mtime that the server responds
+func (f *Fs) Precision() time.Duration {
+	if !f.opt.WaitArchive {
+		return fs.ModTimeNotSupported
+	}
+	return time.Second
```

Comment:
The modtime scheme from `swift.TimeToFloatString` supports Nanosecond precision here?

Reply:
That is for fallback mechanism using `mtime` key where `swift.TimeToFloatString` doesn't take place.

Reply:
Ah, I see. We'd usually assume that all the files have the rclone metadata so have the Nanosecond precision. I don't think it will make a lot of difference in practice though.

Reply:
Makes sense, let's turn it Nanoseconds.